### PR TITLE
WOMBATlite: Treat key biological loss terms semi-implicitly

### PR DESCRIPTION
**Pull request by dougiesquire**
This PR implements the changes described in #96

To summarise, there are a few loss terms in the biological tendency equations that are particularly susceptible to instability with Forward Euler time-stepping:

- phy equation: `phymorq = f(phy^2)`
- zoo equation: `zoomorq = f(zoo^2)`
- det equation: `detremi = f(det^2)`
- caco3 equation: `caldiss = f(caco3)`, `aradiss = f(caco3)`, `pocdiss = f(caco3)`, `zoodiss = f(caco3)`

This PR rewrites these update equations using a semi-implicit scheme for these terms.

Note, I haven't looked at WOMBATmid, but this presumably has similar issues.

### CODE REVIEW

File: generic_tracers/generic_WOMBATlite.F90 [hunk 1]
```diff
@@ -2719,23 +2701,24 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+      ! See https://github.com/ACCESS-NRI/GFDL-generic-tracers/issues/96
+      P_expl = wombat%phygrow(i,j,k) - &
+                wombat%phymorl(i,j,k) - &
+                wombat%zoograzphy(i,j,k)
+      if (biophy>1e-3) then
+        ! Treat phymorq semi-implicitly
+        k_loss = wombat%phyqmor / mmol_m3_to_mol_kg * wombat%f_phy(i,j,k)
+        wombat%f_phy(i,j,k) = (wombat%f_phy(i,j,k) + dtsb * P_expl) / (1.0 + dtsb * k_loss)
+
+        ! Back-calculate phymorq for use below
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90]:
```suggestion
        ! Back-calculate phymorq for use below and for diagnostic output
```

File: generic_tracers/generic_WOMBATlite.F90 [hunk 2]
```diff
@@ -2758,11 +2741,22 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+      ! See https://github.com/ACCESS-NRI/GFDL-generic-tracers/issues/96
+      P_expl = wombat%zooassiphy(i,j,k) + &
+                wombat%zooassidet(i,j,k) - &
+                wombat%zoomorl(i,j,k)
+      if (biozoo>1e-3) then
+        ! Treat zoomorq semi-implicitly
+        k_loss = wombat%zooqmor / mmol_m3_to_mol_kg * wombat%f_zoo(i,j,k)
+        wombat%f_zoo(i,j,k) = (wombat%f_zoo(i,j,k) + dtsb * P_expl) / (1.0 + dtsb * k_loss)
+
+        ! Back-calculate zoomorq for use below
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90]:
```suggestion
        ! Back-calculate zoomorq for use below and diagnostic output
```

File: generic_tracers/generic_WOMBATlite.F90 [hunk 3]
```diff
@@ -2719,23 +2701,24 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+      P_expl = wombat%phygrow(i,j,k) - &
+                wombat%phymorl(i,j,k) - &
+                wombat%zoograzphy(i,j,k)
+      if (biophy>1e-3) then
+        ! Treat phymorq semi-implicitly
+        k_loss = wombat%phyqmor / mmol_m3_to_mol_kg * wombat%f_phy(i,j,k)
+        wombat%f_phy(i,j,k) = (wombat%f_phy(i,j,k) + dtsb * P_expl) / (1.0 + dtsb * k_loss)
+
+        ! Back-calculate phymorq for use below
+        wombat%phymorq(i,j,k) = k_loss * wombat%f_phy(i,j,k) ! [molC/kg/s]
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90:2722]: so now you're calculating `phymorq(i,j,k)` by multiplying `f_phy(i,j,k)` at `timestep = now` and `f_phy(i,j,k)` at `timestep = n+1`? So the actual value of `phymorq(i,j,k)` will be slightly different from what it was before? I guess, come to think of it, that this is exactly what we want since if it was exactly the same as before we suffer instabilities.

File: generic_tracers/generic_WOMBATlite.F90 [hunk 4]
```diff
@@ -2779,13 +2773,34 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
-                              wombat%zooegesdet(i,j,k) + &
-                              wombat%phymorq(i,j,k) + &
-                              wombat%zoomorq(i,j,k) - &
-                              wombat%zoograzdet(i,j,k) - &
-                              wombat%detremi(i,j,k) )
+      ! Here we treat detremi semi-implicitly to ensure numerical stability
+      ! See https://github.com/ACCESS-NRI/GFDL-generic-tracers/issues/96
+      P_expl = wombat%zooegesphy(i,j,k) + &
+                wombat%zooegesdet(i,j,k) + &
+                wombat%phymorq(i,j,k) + &
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90:2785]: so here you're adding the quadratic term that has been calculated semi-implicitly (i.e., slightly different from before). Cool.

File: generic_tracers/generic_WOMBATlite.F90 [hunk 5]
```diff
@@ -2779,13 +2773,34 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+                wombat%zoomorq(i,j,k) - &
+                wombat%zoograzdet(i,j,k)
+      if (wombat%f_det(i,j,k) > epsi) then
+        ! Treat detremi semi-implicitly
+        k_loss = wombat%reminr(i,j,k) / mmol_m3_to_mol_kg * wombat%f_det(i,j,k)
+        wombat%f_det(i,j,k) = (wombat%f_det(i,j,k) + dtsb * P_expl) / (1.0 + dtsb * k_loss)
+
+        ! Back-calculate detremi for use below
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90]:
```suggestion
        ! Back-calculate detremi for use below and diagnostic output
```

File: generic_tracers/generic_WOMBATlite.F90 [hunk 6]
```diff
@@ -2812,12 +2827,31 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+                wombat%zoomorq(i,j,k) * wombat%pic2poc(i,j,k) + &
+                wombat%zoograzphy(i,j,k) * (1. - wombat%fgutdiss) * wombat%pic2poc(i,j,k)
+      if (wombat%f_caco3(i,j,k) > epsi) then
+          ! Treat loss terms that are proportional to CaCO3 concentration implicitly
+          k_loss_zoodiss = wombat%zoograzdet(i,j,k) * wombat%fgutdiss / (biodet * mmol_m3_to_mol_kg)
+          k_loss = wombat%dissratcal(i,j,k) + wombat%dissratara(i,j,k) + wombat%dissratpoc(i,j,k) &
+                        + k_loss_zoodiss
+          wombat%f_caco3(i,j,k) = (wombat%f_caco3(i,j,k) + dtsb * P_expl) / (1.0 + dtsb * k_loss)
+
+          ! Back-calculate the dissolution terms for use below
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90]:
```suggestion
          ! Back-calculate the dissolution terms for use below and diagnostic output
```

File: generic_tracers/generic_WOMBATlite.F90 [hunk 7]
```diff
@@ -2812,12 +2827,31 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
-                                wombat%aradiss(i,j,k) - wombat%pocdiss(i,j,k) )
+      ! Here we treat the loss terms implicitly to ensure numerical stability
+      ! See https://github.com/ACCESS-NRI/GFDL-generic-tracers/issues/96
+      ! Treat the production terms explicitly
+      P_expl = wombat%phymorq(i,j,k) * wombat%pic2poc(i,j,k) + &
+                wombat%zoomorq(i,j,k) * wombat%pic2poc(i,j,k) + &
+                wombat%zoograzphy(i,j,k) * (1. - wombat%fgutdiss) * wombat%pic2poc(i,j,k)
+      if (wombat%f_caco3(i,j,k) > epsi) then
+          ! Treat loss terms that are proportional to CaCO3 concentration implicitly
+          k_loss_zoodiss = wombat%zoograzdet(i,j,k) * wombat%fgutdiss / (biodet * mmol_m3_to_mol_kg)
```
Comment (pearseb) [inline generic_tracers/generic_WOMBATlite.F90:2843]: Isn't this term linear and therefore it isn't necessary to apply a semi-implicit solver?

Above it is:

`wombat%zoograzdet(i,j,k) * wombat%fgutdiss * biocaco3/biodet`

Which is a linear term. It makes sense for the other terms with powers not equal to 1, but this loss term is not a function of any quadratic dependence. For example, you've included `zoograzdet(i,j,k)` above in the other tracer tendency equations with no issue. All we are doing here is doing unit conversions by multiplying it by the CaCO3 to Corg ratio. Are you absolutely sure the implicit solver is necessary here?

Comment (dougiesquire) [reply generic_tracers/generic_WOMBATlite.F90:2843]: @pearseb I needed to rewrite these linear terms (also `caldiss`, `aradiss`, `pocdiss`) as well as the quadratic loss terms to get the OM2 01deg run to progress. Admittedly I started from a state close to the overflow, where concentrations were already potentially quite absurd. I also rewrote all 4 terms at the same time. It's possible we don't need to rewrite all four.

Have you read through my description [here](https://github.com/ACCESS-NRI/GFDL-generic-tracers/issues/96)?

**Comment by dougiesquire** [comment]
@pearseb are you okay with these changes after our discussion today?

I've added your suggestions on the comments. I've also added another commit that guards against some potential overflows - see [a443314](https://github.com/ACCESS-NRI/GFDL-generic-tracers/pull/97/commits/a4433142933e8bcd4985e96dfc45edcb00ca283a)

File: generic_tracers/generic_WOMBATlite.F90 (hunk 7 above)
Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:2843]: Yes I just read it and I am happy.

**Review by pearseb — APPROVED** [review]

**Comment by pearseb** [comment]
Nice :) Great contribution Dougie. ACCESS-OM2-0.1º-WOMBATlite here we come.
